### PR TITLE
target_device with "kind" clause (in metadirective) test

### PR DESCRIPTION
**Pull request by nolanbaker31**
References #637, this is for the "kind" clause specifically.

I am struggling to find any clear list of "isa" or "arch", as these are implementation-defined.

### CODE REVIEW

**Comment by tob2** [comment]
> I am struggling to find any clear list of "isa" or "arch", as these are implementation-defined.

I think there is no way out to either restrict oneself to a small set of values, defined either in the OpenMP spec itself like `kind(host)` or `kind(nohost)` — or those in "OpenMP Additional Definitions" document, both available at https://openmp.org/specifications/

Or to use compiler-specific values, which are hopefully somewhat common on the platforms we are interested it (Nvidia, AMD, Intel?).

_For GCC, those are listed at https://gcc.gnu.org/onlinedocs/libgomp/OpenMP-Context-Selectors.html – the crossrefs go to https://gcc.gnu.org/onlinedocs/gcc/Submodel-Options.html for the respective hardware. I think at least nvidia/gcn values should mostly match LLVM, except that I think LLVM only has 'amdgcn' while GCC had 'gcn' and then added 'amdgcn' as well._

**Review by tob2 — APPROVED** [review]
I have not deeply thought about the testcases and whether testing could be improved (and ignored odd corner cases like default device = host but nonhost devices exist).

But I think the tests should work in sensible environments → LGTM.

**Comment by spophale** [comment]
@mjcarr458 will provide second review.

**Review by spophale — APPROVED** [review]

**Review by jrreap — APPROVED** [review]
Looks good to go!

**Comment by nolanbaker31** [comment]
> Or to use compiler-specific values, which are hopefully somewhat common on the platforms we are interested it (Nvidia, AMD, Intel?).

LLVM has the following list of available "arch" tags: 
`'arm' 'armeb' 'aarch64' 'aarch64_be' 'aarch64_32' 'ppc' 'ppcle' 'ppc64' 'ppc64le' 'x86' 'x86_64' 'amdgcn' 'nvptx' 'nvptx64'`

While x86, x86_64, nvptx and amdgcn are supported, others such as gcn, i386, etc. are not supported in LLVM. I am unsure if writing a test for the shared tags would make sense, as this would not be portable.

While it could be possible to check the compiler on runtime and write the test around this, that could cause its own issues.

File: tests/5.1/metadirective/test_metadirective_target_device_kind.c [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+   int errors = 0;
+   int A[N];
+
+   for (int i = 0; i < N; i++) {
+      A[i] = 1;
+   }
+
+   #pragma omp target map(tofrom: A)
+   {
+      #pragma omp parallel num_threads(4)
```
Comment (nolanbaker31) [inline tests/5.1/metadirective/test_metadirective_target_device_kind.c:28]: Remove num_threads or try num_threads(128) (similar problem to PR #656), test with GCC & NVC

**Comment by spophale** [comment]
Check actual number of threads executing the target region.

**Comment by nolanbaker31** [comment]
Changed to use a similar method to #664 . Seg faults on GCC, but this is because kind is not implemented for target_device yet. (Works with `device={kind(...)}` and `target_device={device_num(...)}`

**Review by spophale — APPROVED** [review]
